File: csharp/triangle/Triangle.cs
using System;

public static class Triangle
{
    public static bool IsScalene(double side1, double side2, double side3)
    {
        return isTrangle(side1, side2, side3)
            && !IsEquilateral(side1, side2, side3)
            && !IsIsosceles(side1, side2, side3);
    }

    public static bool IsIsosceles(double side1, double side2, double side3)
    {
        return isTrangle(side1, side2, side3)
            && (side1 == side2 || side1 == side3 || side2 == side3);
    }

    public static bool IsEquilateral(double side1, double side2, double side3)
    {
        return isTrangle(side1, side2, side3)
                && side1 == side2 && side2 == side3;
    }


    private static bool isTrangle(double side1, double side2, double side3)
    {
        return SideGt0(side1, side2, side3) && AAndBGtC(side1, side2, side3);
    }

    private static bool SideGt0(double side1, double side2, double side3)
    {
        return side1 > 0 && side2 > 0 && side3 > 0;
    }

    private static bool AAndBGtC(double side1, double side2, double side3)
    {
        return side1 + side2 > side3
                && side1 + side3 > side2
                && side2 + side3 > side1;
    }
}

File: zig/triangle/triangle.zig
pub const TriangleError = error{
    Invalid,
};

pub const Triangle = struct {
    // This struct, as well as its fields and methods, needs to be implemented.

    a: f64,
    b: f64,
    c: f64,

    pub fn init(a: f64, b: f64, c: f64) TriangleError!Triangle {
        const ret = Triangle{
            .a = a,
            .b = b,
            .c = c,
        };

        if (!ret.isValid()) {
            return TriangleError.Invalid;
        }

        return ret;
    }

    fn isValid(self: Triangle) bool {
        return self.a > 0 and self.b > 0 and self.c > 0 and self.a + self.b > self.c and self.a + self.c > self.b and self.c + self.b > self.a;
    }

    ///
    /// 是否是等边
    ///
    pub fn isEquilateral(self: Triangle) bool {
        return self.a == self.b and self.a == self.c and self.b == self.c;
    }

    ///
    /// 是否等腰
    ///
    pub fn isIsosceles(self: Triangle) bool {
        return self.a == self.b or self.a == self.c or self.b == self.c;
    }

    ///
    /// 是否是不等边
    ///
    pub fn isScalene(self: Triangle) bool {
        return self.a != self.b and self.a != self.c and self.b != self.c;
    }
};
